magma/compile.py

```diff
@@ -59,7 +59,9 @@ def compile(basename, main, output="coreir-verilog", **kwargs):
     if opts.get("terminate_unused", False):
         TerminateUnusedPass(main).run()
 
-    compiler.compile()
+    result = compiler.compile()
 
     if hasattr(main, "fpga"):
         main.fpga.constraints(basename)
+
+    return result
```

Comment:
Not sure I like the top level compile function returning an arbitrary object that we're right now interpreting as either `None` or a coreir module. Might be pedantic/picky since we pretty much only ever call compile with "coreir" anyway. If we're designing this generally I think a good option would be to return the `compiler` object. Which can store whatever stuff. In this case the coreir compiler can cache the top compiled module. Or keep a log of them.

Reply:
Hmm, but wouldn't returning the result of compilation be a reasonable expectation for a function called `compile`? In most cases, the result is a file (so we could return a file name as a "reference").  Perhaps it makes more sense if we change this to only be behavior when we compile to coreir without producing a file, which would then require this as the only means to retrieve a reference to the compilation result.  (perhaps we can make the output filename an optional argument for this, but raise an error unless it's the coreir backend).

returning the `Compiler` object would be fine as well, but seems a bit odd for a compile function to return a Compiler object to me (especially since it dispatches to Compiler.compile, so I would more expect it to return whatever the reuslt of `Compiler.compile` is)

Reply:
Returning the result of `compile` is definitely a reasonable expectation for a function called `compile` -- but my issue is that depending on the arguments to compile the returned object maybe of completely different type, flavor, and meaning (today it is just `None` or the module, but in general this feels weird). I also agree returning the `Compiler` object is weird.

I think I would feel better if we *always* returned an attribute dict, which depending on the arguments to `compile` will have known keys. It can definitely be empty. In the CoreIR case it should say return `{"coreir_module": ...}`.

Let me propose the change I'm thinking.

Reply:
This seems reasonable, and I could see adding other things like "file_name" or other useful information like a file list